buildscript-utils/src/main/kotlin/com/fueledbycaffeine/spotlight/buildscript/SpotlightProjectList.kt

```diff
@@ -1,48 +1,121 @@
+
   public companion object {
     private const val COMMENT_CHAR = "#"
     public const val ALL_PROJECTS_LOCATION: String = "gradle/all-projects.txt"
     public const val IDE_PROJECTS_LOCATION: String = "gradle/ide-projects.txt"
 
     @JvmStatic
-    public fun allProjects(buildRoot: Path): SpotlightProjectList =
-      SpotlightProjectList(buildRoot, buildRoot.resolve(ALL_PROJECTS_LOCATION))
+    public fun allProjects(buildRoot: Path): AllProjects =
+      AllProjects(buildRoot, buildRoot.resolve(ALL_PROJECTS_LOCATION))
 
     @JvmStatic
-    public fun ideProjects(buildRoot: Path): SpotlightProjectList =
-      SpotlightProjectList(buildRoot, buildRoot.resolve(IDE_PROJECTS_LOCATION))
+    public fun ideProjects(buildRoot: Path, allProjects: (() -> Set<GradlePath>)? = null): IdeProjects =
+      IdeProjects(buildRoot, buildRoot.resolve(IDE_PROJECTS_LOCATION), allProjects)
   }
 
-  public fun read(): Set<GradlePath> {
-    if (projectList.notExists()) return emptySet()
+  public fun read(): Set<GradlePath>
+
+  public infix fun contains(path: GradlePath): Boolean = read().contains(path)
+
+  public fun ensureFileExists() {
+    if (projectList.notExists()) projectList.createFile()
+  }
 
+  public fun readRawPaths(includeComments: Boolean): List<String> {
+    if (projectList.notExists()) return emptyList()
     return projectList.readLines()
-      .filterNot { line -> line.trim().startsWith(COMMENT_CHAR) || line.isBlank() }
+      .filterNot { line ->
+        (!includeComments && line.trim().startsWith(COMMENT_CHAR)) || line.isBlank()
+      }
+  }
+}
+
+public class AllProjects internal constructor(
+  override val buildRoot: Path,
+  override val projectList: Path
+) : SpotlightProjectList {
+
+  override fun read(): Set<GradlePath> {
+    return readRawPaths(includeComments = false)
       .map { GradlePath(buildRoot, it) }
       .toSet()
   }
+}
 
-  public infix fun contains(path: GradlePath): Boolean = read().contains(path)
+public class IdeProjects internal constructor(
+  override val buildRoot: Path,
+  override val projectList: Path,
+  private val allProjects: (() -> Set<GradlePath>)? = null
+) : SpotlightProjectList {
 
-  public fun ensureFileExists() {
-    if (projectList.notExists()) projectList.createFile()
+  private companion object {
+    private const val STAR_CHAR = '*'
+    private const val QUESTION_CHAR = '?'
+
+    private fun String.containsGlobChar(): Boolean {
+      for (char in this) {
+        if (char == STAR_CHAR || char == QUESTION_CHAR) {
+          return true
+        }
+      }
+      return false
+    }
   }
 
+  // TODO filter out paths already covered by a glob
   public fun add(paths: Iterable<GradlePath>) {
     ensureFileExists()
-    projectList.writeText((read() + paths).joinToString("\n") { it.path })
+    projectList.writeText((readRawPaths(includeComments = true) + paths.map { it.path }).joinToString("\n"))
   }
 
+  // TODO detect if paths are included by a pattern and signal to user
```

Comment:
Left these two in here for your thoughts, though think they might be better left to a subsequent PR since they might require some UI in the IDE plugin